### ISSUE TITLE
Spelling / Grammar checking

### ISSUE THREAD

**Issue by exabrial** [gollum/gollum#979]
It would superb if the wiki markdown editor had a toolbar that supported spelling and grammar checking. There are many Javascript and Ruby libraries available to do this on Github with various licenses available so it could be implemented either client or server side.

The stopgap measure we've come up with is to toss the `.md` documents in Word, but Word hates markdown and barfs at the formatting. 

If someone can give us a little direction on this: How the Gollum authors would like it implemented, where to put it, etc we could at least take a crack at an initial implementation.

**Comment by sunny** [comment]
Hey @exabrial! That would be cool.

My 2 cents:

- it would be nice to support more than english
- the ACE editor seems to have [an extension for that](https://github.com/swenson/ace_spell_check_js)  but we might not keep ACE in the future…
- browsers tend to add their own 
![spelling correcter](https://cloud.githubusercontent.com/assets/132/6418431/522209fa-beb5-11e4-8747-4e9c7958dfd2.png) which makes a lot of sense considering that this is a feature that should be included  on every website in every language and requires a lot of data…

**Comment by dometto** [comment]
For discussions on ACE see #826 -- I'm not opposed to keeping it per se, as long as we can find a good way to simplify maintenance.

**Comment by dometto** [comment]
I wonder whether this is still a feature we need, given the existence of in-browser spellchecking?

**Comment by bartkamphorst** [comment]
I personally don't use it, but wouldn't mind it if it were optional and available for many languages.

**Comment by dometto** [comment]
Looks like in 5.x we don't have any spell checking, also not browser-based, because apparently doesn't play well with e.g. Firefox's spellchecking ability. @ProgramFan, do you have any experience getting spellchecking to work in ace? All I found is this old-looking repo: https://github.com/swenson/ace_spell_check_js

**Comment by ProgramFan** [comment]
I have tried to implement spell checking in Ace and no success yet. I have tried the following two approaches:
1. Enable browser checking by setting `spellcheck="true"` for the editor div.
2. Use the bundled ext-spell-check plugin of Ace.

I am still looking ways around it.

**Comment by dometto** [comment]
Do we still want this in `5.x`? @ProgramFan do you happen to know if this has gotten any better with recent releases of ACE?

**Comment by ProgramFan** [comment]
Unfortunately, this issue https://github.com/ajaxorg/ace/issues/2107 is still opening. And I don't have enough time to investigate the inners of ace. So maybe we have to abandon either ACE editor or spell checking.

**Comment by dometto** [comment]
I've got Firefox's spellchecking to work with ACE. Adding `contenteditable="true" spellcheck="true"` to the editor-div produces this result:

<img width="1059" alt="screenshot 2018-11-29 at 23 59 41" src="https://user-images.githubusercontent.com/147111/49258041-51a45100-f434-11e8-8e1d-48c1b0378441.png">


As you can see it's not fully syntax aware, but it's helpful for the prose bits -- and we could turn it off by default, providing the user with a 'check spelling' button which sets the relevant attributes to true. Bad news: doesn't seem to work in Chrome or Safari. Any input welcome.

**Comment by bartkamphorst** [comment]
I think this is the best we can do for now, unless someone develops a functional ACE plugin. The `contenteditable` and `spellcheck` are non-normative according to the HTML5 spec, but I wouldn't be surprised if other browsers will start to adhere to them in the future. In any case, it makes sense to me that the browser handles spell checking in a uniform way across websites.

**Comment by dometto** [comment]
Agreed. For now, do you think we should add a button? If so, should it be disabled unless the browser is FF?

**Comment by bartkamphorst** [comment]
> Adding contenteditable="true" spellcheck="true" to the editor-div produces this [...]

@dometto Which div exactly? Because if I read [this](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/spellcheck) correctly, these attributes will work in all browsers if they are added to the `textarea` instead of the div.

**Comment by dometto** [comment]
I injected (using the Web Developer tools in FF) the attributes into the `gollum-editor` div, which is created by ace (i.e. it doesn't appear in the mustache template because it's generated by the JS). The attributes for the `gollum-editor` textarea (which is in the mustache template) seem to be overridden by ace, so adding them there is no good. Injecting the attributes directly into the html for the textarea appears to have no effect.

**Comment by dometto** [comment]
I've noticed, though, that the textarea that appears on the loaded page (so the one that's been modified by ace's JS) already has it's spellcheck attribute set to true. That is the work of the `ace-spellcheck` extension (if I remove that from `app.js`, spellcheck is set to false). The purpose of `ace-spellcheck` is to show the normal in-browser context menu when the user right-clicks on the editor (rather than showing ace's internal dialog), so that spellchecking can work. But it doesn't seem sufficient because it doesn't also set `contenteditable=true`.

**Comment by bartkamphorst** [comment]
Does injecting `contenteditable=true` to the textarea after the DOM has been loaded have any effect?

**Comment by dometto** [comment]
Weirdly, no. I've tested with updating the attributes via JS (in the console) now, and the only thing that seems to work is:
```js
$('div#gollum-editor').attr('contenteditable', 'true'); $('div#gollum-editor').attr('spellcheck', 'true')
```
...and that only in FF.

**Comment by bartkamphorst** [comment]
Interesting. Let's keep digging, because the simple textarea [here](https://www.wufoo.com/html5/spellcheck-attribute/) does provide spellchecking in Safari.

**Comment by dometto** [comment]
Yes, we rely quite heavily on ace to do things like implement custom syntax coloring (e.g. for macros, code blocks), and to determine the location of subsections in a document (to implement the 'editable sections' feature). Monaco looks very good though, also the fact that it has a diff editor -- but we would have to port quite a lot of features to the new editor. It's frustrating that spellchecking is apparently so difficult to support in ace.
